### ISSUE TITLE
Change Request: Allow `SourceCode#getText()` types to accept tokens and comments

### ISSUE THREAD

**Issue by electrohyun** [eslint/eslint#21329]
### ESLint version

v10.10.0

### What problem do you want to solve?

I found this while looking through [a rule file](https://github.com/eslint-community/eslint-plugin-n/blob/6a9980a999defd64664a3b9626b6a9f2f67fac67/lib/rules/exports-style.js#L198-L199) in the `eslint-community`. The file uses `@ts-expect-error` when passing a comment to `SourceCode#getText()`.

The type definition of `SourceCode#getText()` only accepts an `ESTree.Node`.

```ts
getText(
    node?: ESTree.Node,
    beforeCount?: number,
    afterCount?: number,
): string;
```

Other `SourceCode` methods return `AST.Token` and `ESTree.Comment` values.

```ts
const token = sourceCode.getFirstToken(node);
// AST.Token | null

const comments = sourceCode.getCommentsBefore(node);
// ESTree.Comment[]
```

Despite this type definition, `getText()` can also accept these values because its implementation only reads the provided object's `range`.

```js
getText(node, beforeCount, afterCount) {
    if (node) {
        return this.text.slice(
            Math.max(node.range[0] - (beforeCount || 0), 0),
            node.range[1] + (afterCount || 0),
        );
    }

    return this.text;
}
```

So, this causes valid runtime code to fail type checking.

```ts
const token = sourceCode.getFirstToken(node);

if (token) {
  // Argument of type 'Token' is not assignable to parameter of type 'Node | undefined'.
  sourceCode.getText(token);
}

for (const comment of sourceCode.getCommentsBefore(node)) {
  // Argument of type 'Comment' is not assignable to parameter of type 'Node | undefined'.
  sourceCode.getText(comment);
}
```

This means that values returned by `SourceCode` itself cannot be passed back to `getText()` without a type error, even though they are handled at runtime.

<hr />

I also found examples of this usage in ESLint.

- [The `curly` rule](https://github.com/eslint/eslint/blob/v10.10.0/lib/rules/curly.js#L298-L304) passes a token returned by `getTokenBefore()` to `getText()`.
- [A `SourceCode` test](https://github.com/eslint/eslint/blob/v10.10.0/tests/lib/languages/js/source-code/source-code.js#L373-L383) passes a shebang comment returned by `getAllComments()` to `getText()`.

And found similar usages in ecosystem plugins.

- [`eslint-plugin-n`](https://github.com/eslint-community/eslint-plugin-n/blob/6a9980a999defd64664a3b9626b6a9f2f67fac67/lib/rules/exports-style.js#L198-L199) passes a comment to `getText()` and uses `@ts-expect-error` to suppress the resulting type error.
- [`eslint-plugin-typescript-sort-keys`](https://github.com/infctr/eslint-plugin-typescript-sort-keys/blob/d32e4dada117f2369e4b07e72f516edbb605d557/src/utils/plugin.ts#L116-L122) passes a comment to `getText()` using `comment as any`. removing the type assertion reproduces the same error.
- [`eslint-plugin-es`](https://github.com/mysticatea/eslint-plugin-es/blob/b112d4dea8da6c27cfb2d086d085f71139f2d41a/lib/rules/no-optional-chaining.js#L31-L37) passed a token to `getText()` as part of an Espree compatibility workaround.
- [`eslint-stylistic`](https://github.com/eslint-stylistic/eslint-stylistic/blob/7736997ad57cf9c6e4b324c37534cfb47d187003/packages/eslint-plugin/rules/keyword-spacing/keyword-spacing.ts#L93-L101) also passes a token to `getText()` (without a type error because it uses `@typescript-eslint`'s `SourceCode` type that already accepts `Node | Token`).

### What do you think is the correct solution?

I think we can update `SourceCode#getText()` to accept nodes, tokens, and comments.

```ts
getText(
    syntaxElement?: ESTree.Node | AST.Token | ESTree.Comment,
    beforeCount?: number,
    afterCount?: number,
): string;
```

This would be a type widening and would not require a runtime implementation change.

Type tests can cover all three supported inputs.

```ts
sourceCode.getText(AST);
sourceCode.getText(TOKEN);
sourceCode.getText(COMMENT);
```

The documentation and implementation JSDoc can also be updated to describe the parameter as a node, token, or comment.

### Participation

- [x] I am willing to submit a pull request for this change.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

<hr />

Disclosure: I'm a participant of [open source contribution program OSSCA](https://github.com/eslint-ossca)

**Comment by fasttime** [comment]
Thanks for the detailed report @electrohyun. I support this proposal, since it would align the types with the actual usage. The types themselves should be easy to update, but I think it would be necessary to also update the documentation parts where `getText()` is indicated as accepting a _node_ argument, both in this repo and in `eslint/rewrite`, particularly in [`@eslint/plugin-kit`](https://github.com/eslint/rewrite/blob/f83339de1a393800eaa3c07c1266582e986c06b7/packages/plugin-kit/src/source-code.js#L606-L616), and including JSDoc comments.

I would like to get opinions from @eslint/eslint-tsc, since this change would affect usage of the public ESLint API, even though the behavior would remain unchanged.

**Comment by nzakas** [comment]
Yes, I think we can update the types. Really, `getText()` can accept anything that has location information on it because it just uses the range to slice the string. That's why `@eslint/plugin-kit` specifies the first argument as `SyntaxElementWithLoc`. 

I think we can just delete these lines and we'll inherit the correct types:
https://github.com/eslint/eslint/blob/9ac7eb60525ec768c2fdc6699e5a292aa913ce9e/lib/types/index.d.ts#L262-L266

Then update this line:
https://github.com/eslint/eslint/blob/9ac7eb60525ec768c2fdc6699e5a292aa913ce9e/lib/types/index.d.ts#L243

Updating docs is always a good idea.

**Comment by fasttime** [comment]
> I think we can just delete these lines and we'll inherit the correct types:
> 
> https://github.com/eslint/eslint/blob/9ac7eb60525ec768c2fdc6699e5a292aa913ce9e/lib/types/index.d.ts#L262-L266
> 
> Then update this line:
> 
> https://github.com/eslint/eslint/blob/9ac7eb60525ec768c2fdc6699e5a292aa913ce9e/lib/types/index.d.ts#L243

Makes sense. We can update the `SourceCode` type to extend `TextSourceCodeBase` from `@eslint/plugin` kit:

```diff
- export class SourceCode implements TextSourceCode<{
+ export class SourceCode extends TextSourceCodeBase<{
```

That should make `getText()` and other member definitions redundant. Even if the JavaScript `SourceCode` class does not inherit from `TextSourceCodeBase`, the types should be largely compatible.

**Comment by electrohyun** [comment]
@nzakas @fasttime Thanks for the feedback. I'll work on it this way.

But there's one thing about `getParent()`. It's now available in the types, but the `SourceCode` class doesn't implement it. 

I think we could add a method that returns `node.parent` or `undefined` if there's no parent, like this.

```js
getParent(node) {
    return node.parent ?? undefined;
}
```

Would this approach make sense? I'd appreciate your thoughts.

**Comment by fasttime** [comment]
@electrohyun thanks for looking into this. I'd suggest adding `getParent: never` to the types until the differences between JS and the other languages are leveled out.

**Comment by electrohyun** [comment]
Hello. For consistency with the disclosure in PR, I’ve updated the AI acknowledgment in this issue to note that I used AI only to refine the English wording of the report I wrote myself.

Thank you again, and my apologies for the additional comment.
